algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs

```diff
@@ -106,16 +125,23 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
                 prover::SingleEntry { z_a, z_b, w_poly, z_a_poly, z_b_poly }
             })
             .collect::<Vec<_>>();
-        assert_eq!(batches.len(), batch_size);
-
-        let mask_poly = Self::calculate_mask_poly(constraint_domain, rng);
-
-        let oracles = prover::FirstOracles { batches, mask_poly };
-        assert!(oracles.matches_info(&Self::first_round_polynomial_info(batch_size)));
+        assert_eq!(batches.len(), state.total_instances);
+
+        let mut batch_consumed_so_far = 0;
+        let mut circuit_specific_batches = BTreeMap::new();
+        for ((circuit, state), r_b_s) in state.circuit_specific_states.iter_mut().zip(r_b_s) {
+            let batches = batches[batch_consumed_so_far..][..state.batch_size].to_vec();
```

Comment:
Does this perform a clone of large values?

Reply:
It did, my bad, draining the batches vector now: b17840352